contents/javascript/Javascript-modules.md

```diff
@@ -0,0 +1,171 @@
+<frontmatter>
+  title: "Javascript: Modules"
+  footer: footer.md
+  head: head.md
+  siteNav: mainNav.md
+  pageNav: 3
+</frontmatter>
+
+{{ navbar | safe }}
+
+<div class="website-content">
+
+# Javascript: Modules
+
+Author: Gilbert Emerson
+
+## What Is a Module?
+
+You might often encounter the term such as module, package, library, dependency, plugin, etc. We will not dive into the exact definition of each term (because that is not the scope of this chapter), but we will loosely define module as a collection of code that is placed in different file for various reasons.
+
+We would definitely want to modularize our code for good reason, which will be expanded on the section below.
+
+## Why Use Modules?
+
+There can be a lot of reason why we would like to modularize our code, but here are 3 of the more important one:
+
+1. Maintainability
+A well designed module will be self-contained. Familiar principles of Software Engineering such as SOLID, etc can be applied to design a good module. A good module will allow a better maintanability of the module.
+
+2. Namespacing
+Because of how JavaScript scoping works (not convered in this article), namespacing can be quite a big issue in JavaScript because of global namespace pollution. Modules allows JavaScript engineer to avoid this problem by isolating the namespace to a specific module.
+
+3. Reusability
+As per DRY Principle, module allows developers to reuse their code that is contained in a module.
+
+## How Is It Implemented?
+
+Historically, JavaScript did not support modules up until 2015 with the inception of EcmaScript6 (ES6). Nevertheless, there have been "unofficial implementations" of module sytem in JavaScript in form of Module Pattern and CommonJS.
+
+This article will expand on those 3 implementations.
+
+### Module Pattern
+
+This is the way JavaScript developer create module in the past before the inception of CommonJS and ES6. Using a technique in JavaScript called IIFE (Immediately-Invoked Function Expression), they wrap their code in an IIFE.
+
+> IIFE syntax </br>
+> `(function() {})()`
+
+`anExampleModule.js`
+```javascript
+var anExampleModule = (function() {
+    var variableOne = 1;
+    var variableTwo = 2;
+
+    var getVariableOne = function() {
+        return variableOne;
+    }
+
+    var sumOfVariable = function() {
+        return variableOne + variableTwo;
+    }
+
+    var counter = function() {
+        variableOne++;
+    }
+
+    return {
+        getVariableOne: getVariableOne,
+        sumOfVariable: sumOfVariable,
+        counter,
+    }
+})()
+```
+
+`index.js`
+```javascript
+anExampleModule.getVariableOne(); // 1
+anExampleModule.sumOfVariable(); // 3
+```
+
+`index.html`
+```html
+<script src="./anExampleModule.js" />
+<script src="./index.js" />
+```
```

Comment:
actually splitting how the module is used into a separate file, doesn't really help with the explanation. 

maybe just put it in the html? with a comment that now `anExampleModule` is global?